include/rovio/MultilevelPatchAlignment.hpp

```diff
@@ -129,12 +202,93 @@ class MultilevelPatchAlignment {
-    A.col(1).array() -= mean_diff_dy;
+
+    float reg_a, reg_a_dx, reg_a_dy, reg_b, reg_b_dx, reg_b_dy;
+    if(useIntensityOffset_ || useIntensitySqew_){
+      mean_x = mean_x/wTot;
+      mean_xx = mean_xx/wTot;
+      mean_xy = mean_xy/wTot;
+      mean_xy_dx = mean_xy_dx/wTot;
+      mean_xy_dy = mean_xy_dy/wTot;
+      mean_y = mean_y/wTot;
+      mean_y_dx = mean_y_dx/wTot;
+      mean_y_dy = mean_y_dy/wTot;
+
+      if(useIntensitySqew_){
+        reg_a = (mean_xy-mean_x*mean_y)/(mean_xx-mean_x*mean_x);
+        reg_a_dx = (mean_xy_dx-mean_x*mean_y_dx)/(mean_xx-mean_x*mean_x);
+        reg_a_dy = (mean_xy_dy-mean_x*mean_y_dy)/(mean_xx-mean_x*mean_x);
+        if(reg_a < 0.5f){
+          reg_a = 0.5;
+          reg_a_dx = 0.0;
+          reg_a_dy = 0.0;
+        }
+      } else {
+        reg_a = 1.0;
+        reg_a_dx = 0.0;
+        reg_a_dy = 0.0;
+      }
+      if(useIntensityOffset_){
+        reg_b = mean_y-reg_a*mean_x;
+        reg_b_dx = mean_y_dx-reg_a_dx*mean_x;
+        reg_b_dy = mean_y_dy-reg_a_dy*mean_x;
+      } else {
+        reg_b = 0.0;
+        reg_b_dx = 0.0;
+        reg_b_dy = 0.0;
+      }
+    }
+
+    // Compute correct patch error and gradient (based on linear brightness fix), apply Huber norm, apply weighting
+    numLevel = 0;
+    for(int l = l1; l <= l2; l++){
+      if(mlpError_.isValidPatch_[l]){
+        numLevel++;
+        A.conservativeResize(numLevel*patch_size*patch_size,2);
+        b.conservativeResize(numLevel*patch_size*patch_size,1);
+        const float* it_patch = mp.patches_[l].patch_;
+        const float* it_patch_extracted = extractedPatches_[l].patch_;
+        float* it_error = mlpError_.patches_[l].patch_;
+        float* it_dx_error = mlpError_.patches_[l].dx_;
+        float* it_dy_error = mlpError_.patches_[l].dy_;
+        const float* it_w = &w_[l*patch_size*patch_size];
+        for(int y=0; y<patch_size; ++y){
+          for(int x=0; x<patch_size; ++x, ++it_patch, ++it_patch_extracted, ++it_error,  ++it_dx_error, ++it_dy_error, ++it_w){
+            if(useIntensityOffset_ || useIntensitySqew_){
+              *it_error = *it_patch_extracted - reg_a*(*it_patch) - reg_b;
+              *it_dx_error = reg_a*(*it_dx_error - reg_a_dx*(*it_patch) - reg_b_dx);
+              *it_dy_error = reg_a*(*it_dy_error - reg_a_dy*(*it_patch) - reg_b_dy);
+            }
+            b((numLevel-1)*patch_size*patch_size+y*patch_size+x,0) = *it_error;
+            A((numLevel-1)*patch_size*patch_size+y*patch_size+x,0) = *it_dx_error;
+            A((numLevel-1)*patch_size*patch_size+y*patch_size+x,1) = *it_dy_error;
+
+            if(huberNormThreshold_ > 0.0){ // TODO: investigate why 1.0 leads to non-deterministic behavior
+              const float b_abs = std::fabs(*it_error);
+              if(b_abs > huberNormThreshold_){
+                b((numLevel-1)*patch_size*patch_size+y*patch_size+x,0) = std::sqrt(huberNormThreshold_*(2.0*b_abs - huberNormThreshold_));
```

Comment:
I think you need to add the sign
`b((numLevel-1)*patch_size*patch_size+y*patch_size+x,0) = std::sqrt(huberNormThreshold_*(2.0*b_abs - huberNormThreshold_))*copysign(1.0, *it_error);`